Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java

```diff
@@ -28,44 +28,224 @@
+     */
+    abstract void copyPoint(double[] point, int src, int location, int length);
+
+    /**
+     * Add a point to the point store and return the index of the stored point.
+     *
+     * @param point The point being added to the store.
+     * @return the index value of the stored point.
+     * @throws IllegalArgumentException if the length of the point does not match
+     *                                  the point store's dimensions.
+     * @throws IllegalStateException    if the point store is full.
      */
-    public int getRefCount(int index) {
-        return refCount[index];
+    public int add(double[] point) {
+        checkArgument(point.length == dimensions, "point.length must be equal to dimensions");
+
+        if (shingleSize > 1 && shingleAwareOverlapping) {
+            /**
+             * corresponds to the shingled/overlapped case. Currently we are considering a
+             * perfect sequence where the shift corresponds to one time unit and that
+             * represents baseDimension new values. Note that if the alignment test fails,
+             * then the points are still copied -- in future it may be reasonable to check
+             * for different alignments -- which would be beneficial for larger shingles. In
+             * extreme a segment tree type data structure could also be used.
+             */
+            if (startOfFreeSegment > capacity * dimensions - baseDimension) {
+                // the above ensures that the most recent values in a shingle can be writtem
+                compact();
+            }
+
+            boolean test = (startOfFreeSegment - dimensions + baseDimension >= 0);
+            if (test) {
+                test = checkShingleAlignment(startOfFreeSegment, point);
+            }
+            if (test && startOfFreeSegment + baseDimension <= capacity * dimensions) {
+                int nextIndex = takeIndex(); // no more compactions
+                refCount[nextIndex] = 1; // has to be after compactions
+                locationList[nextIndex] = startOfFreeSegment - dimensions + baseDimension;
+                copyPoint(point, dimensions - baseDimension, startOfFreeSegment, baseDimension);
+                startOfFreeSegment += baseDimension;
+                return nextIndex;
+            }
+
+            // we must write the full contents of the point; we need to check for space
+            if (startOfFreeSegment >= capacity * dimensions - dimensions) {
+                compact();
+            }
+            int nextIndex = takeIndex(); // no more compactions
+            locationList[nextIndex] = startOfFreeSegment;
+            copyPoint(point, 0, startOfFreeSegment, dimensions);
+            startOfFreeSegment += dimensions;
+            refCount[nextIndex] = 1; // has to be after compactions
+            return nextIndex;
+        }
+        /**
+         * the following corresponds to shingleSize=1 or the no-overlap case. The
```

Comment:
This is more a matter of personal taste, but I might implement this method as
```
if (shingleSize > 1 && shingleAwareOverlapping) {
  addShingleAwareOverlapping(point);
} else {
  addSimple(point); // or addDirect maybe?
}
```
where `addShingleAwareOverlapping` and `addSimple` are private or protected methods.